File: src/engine/policy-engine.ts
import { SealIssue, SealInput, LLMSignals, Verdict, RiskLevel, maxVerdict, makeIssue } from '../types.ts'

const DESTRUCTIVE_PATTERN = /\b(DROP\s+TABLE|rm\s+-rf|DELETE\s+FROM|TRUNCATE|format\s+\/|fdisk)\b/i
const PRODUCTION_DEPLOY_PATTERN = /\b(deploy(?:ing)?|release|push to production|go live)\b/i
const MIGRATION_PATTERN = /\b(migration|migrate|schema\s+change)\b/i
const ROLLBACK_PATTERN = /\b(rollback|revert|down\s+migration|undo)\b/i

export interface PolicyResult {
  verdict: Verdict
  injected_issues: SealIssue[]
  policy_deductions: number
}

export class PolicyEngine {
  static apply(params: {
    base_verdict: Verdict
    detector_score: number
    risk_level: RiskLevel
    input: SealInput
    all_findings: SealIssue[]
    llm_signals?: LLMSignals | null
  }): PolicyResult {
    const { base_verdict, input, risk_level, all_findings } = params
    const injected: SealIssue[] = []
    let verdict: Verdict = base_verdict
    let policy_deductions = 0

    // AX105: HIGH/CRITICAL risk without evidence
    const hasMissingEvidence = all_findings.some(f => f.type === 'MISSING_EVIDENCE' && f.is_blocking)
    const riskOrder = ['LOW', 'MEDIUM', 'HIGH', 'CRITICAL']
    if (riskOrder.indexOf(risk_level) >= riskOrder.indexOf('HIGH') && hasMissingEvidence) {
      injected.push(makeIssue({
        type: 'SECURITY_RISK', severity: 'HIGH', layer: 'L4', source: 'core',
        rule_id: 'AX105',
        evidence: `${risk_level} risk artifact has missing evidence`,
        required_fix: 'Provide evidence for all claims in high-risk artifact',
      }))
      verdict = maxVerdict(verdict, 'REVISE')
    }

    // AX501: destructive commands without confirmation
    if (DESTRUCTIVE_PATTERN.test(input.output) && !/human confirmation|approve|confirm/i.test(input.output)) {
      injected.push(makeIssue({
        type: 'DATA_RISK', severity: 'CRITICAL', layer: 'L4', source: 'core',
        rule_id: 'AX501',
        evidence: 'Destructive command recommended without human confirmation instruction',
        required_fix: 'Add explicit human confirmation requirement before executing destructive command',
      }))
      verdict = 'BLOCK'
    }

    // AX502: production deployment for HIGH/CRITICAL
    if (PRODUCTION_DEPLOY_PATTERN.test(input.output) && riskOrder.indexOf(risk_level) >= riskOrder.indexOf('HIGH')) {
      injected.push(makeIssue({
        type: 'SECURITY_RISK', severity: 'HIGH', layer: 'L4', source: 'core',
        rule_id: 'AX502',
        evidence: `Production deployment recommended for ${risk_level} risk artifact`,
        required_fix: 'Require human approval before production deployment',
      }))
      verdict = maxVerdict(verdict, 'ESCALATE_TO_HUMAN')
    }

    // AX503: migration without rollback
    const isMigration = input.artifact_type === 'migration' || MIGRATION_PATTERN.test(input.output)
    const hasRollback = ROLLBACK_PATTERN.test(input.output) || ROLLBACK_PATTERN.test(input.evidence.diff)
    if (isMigration && !hasRollback) {
      injected.push(makeIssue({
        type: 'DATA_RISK', severity: 'CRITICAL', layer: 'L4', source: 'core',
        rule_id: 'AX503', trust_deduction: 25,
        evidence: 'Migration detected but no rollback plan found',
        required_fix: 'Add rollback/revert plan to migration',
      }))
      policy_deductions += 25
      verdict = maxVerdict(verdict, 'REVISE')
    }

    // AX504: security change without security tests (TW205 fired)
    const tw205Fired = all_findings.some(f => f.rule_id === 'TW205')
    if (tw205Fired) {
      verdict = maxVerdict(verdict, 'ESCALATE_TO_HUMAN')
    }

    // CL402 enforcement
    const cl402Fired = all_findings.some(f => f.rule_id === 'CL402')
    if (cl402Fired) verdict = maxVerdict(verdict, 'ESCALATE_TO_HUMAN')

    // LLM signal conversion
    let llm_policy_deductions = 0
    if (params.llm_signals) {
      const signals = params.llm_signals
      if (signals.confidence < 0.6 && riskOrder.indexOf(risk_level) >= riskOrder.indexOf('MEDIUM')) {
        injected.push(makeIssue({
          type: 'AMBIGUITY', severity: 'MEDIUM', layer: 'L2', source: 'llm-overlay',
          rule_id: 'CL403', trust_deduction: 20,
          evidence: `LLM reviewer confidence ${signals.confidence.toFixed(2)} below threshold for ${risk_level} risk`,
        }))
        llm_policy_deductions += 20
        verdict = maxVerdict(verdict, 'ESCALATE_TO_HUMAN')
      }
    }

    // required_verdict overrides from extension/detector issues
    for (const issue of all_findings) {
      if (issue.required_verdict) {
        verdict = maxVerdict(verdict, issue.required_verdict)
      }
    }

    // BLOCK is terminal
    if (verdict === 'BLOCK') {
      // stays BLOCK
    }

    return { verdict, injected_issues: injected, policy_deductions: policy_deductions + llm_policy_deductions }
  }

  static resolveVerdicts(verdicts: Verdict[]): { verdict: Verdict } {
    const order: Verdict[] = ['PASS', 'PASS_WITH_WARNINGS', 'REVISE', 'ESCALATE_TO_HUMAN', 'BLOCK']
    const maxIdx = Math.max(...verdicts.map(v => order.indexOf(v)))
    return { verdict: order[maxIdx] ?? 'PASS' }
  }

  static convertLLMSignals(signals: LLMSignals): SealIssue[] {
    const issues: SealIssue[] = []
    const map = [
      { field: 'suspected_issues' as const, type: 'LOGIC_BUG' as const, severity: 'MEDIUM' as const },
      { field: 'missing_requirements' as const, type: 'SPEC_MISMATCH' as const, severity: 'HIGH' as const },
      { field: 'evidence_gaps' as const, type: 'MISSING_EVIDENCE' as const, severity: 'MEDIUM' as const },
      { field: 'possible_edge_cases' as const, type: 'AMBIGUITY' as const, severity: 'LOW' as const },
    ]
    for (const { field, type, severity } of map) {
      for (const text of signals[field]) {
        issues.push(makeIssue({ type, severity, layer: 'L2', source: 'llm-overlay', evidence: text }))
      }
    }
    return issues
  }
}

File: python/seal_gate/engine/policy_engine.py
import re
from ..types import SealIssue, SealInput, make_issue, max_verdict, RISK_ORDER

# AX501: no trailing \b after non-word char /
DESTRUCTIVE = re.compile(r'\b(DROP\s+TABLE|rm\s+-rf|DELETE\s+FROM|TRUNCATE|format\s+/|fdisk)', re.I)
CONFIRMATION_NEAR = re.compile(r'\b(requires?\s+(?:human\s+)?(?:confirmation|approval|review)|please\s+confirm|must\s+(?:approve|confirm))\b', re.I)
# AX502: only explicit deployment actions, not "release notes"
PRODUCTION_DEPLOY = re.compile(r'\b(deploy(?:ing)?\s+to\s+(?:prod|production)|push\s+to\s+production|go\s+live|release\s+to\s+(?:prod|production))\b', re.I)
MIGRATION_RE = re.compile(r'\b(migration|migrate|schema\s+change)\b', re.I)
# AX503: require explicit rollback plan, not just "git revert" or "cannot undo"
ROLLBACK_RE = re.compile(r'\b(rollback\s+(?:plan|script|procedure|step)|revert\s+(?:migration|schema|change)|down\s+migration|migration\s+rollback)\b', re.I)

LLM_SIGNAL_MAP = {
    'suspected_issues':    ('LOGIC_BUG',       'MEDIUM'),
    'missing_requirements':('SPEC_MISMATCH',    'HIGH'),
    'evidence_gaps':       ('MISSING_EVIDENCE', 'MEDIUM'),
    'possible_edge_cases': ('AMBIGUITY',        'LOW'),
}


def _risk_idx(r: str) -> int:
    try:
        return RISK_ORDER.index(r)
    except ValueError:
        return len(RISK_ORDER) - 1  # unknown → max risk (fail-safe)


def _has_adjacent_confirmation(output: str, match_start: int) -> bool:
    window = output[max(0, match_start - 150):match_start + 150]
    return bool(CONFIRMATION_NEAR.search(window))


def apply_policy(base_verdict: str, input: SealInput, risk_level: str, all_findings: list[SealIssue], llm_signals: dict | None = None) -> dict:
    injected: list[SealIssue] = []
    verdict = base_verdict
    policy_deductions = 0

    # Validate risk_level — unknown → treat as CRITICAL (fail-safe)
    risk_level_safe = risk_level if risk_level in RISK_ORDER else 'CRITICAL'

    # AX105: check severity directly, not brittle is_blocking flag
    has_missing_evidence = any(
        f.type == 'MISSING_EVIDENCE' and f.severity in ('CRITICAL', 'HIGH')
        for f in all_findings
    )
    if _risk_idx(risk_level_safe) >= _risk_idx('HIGH') and has_missing_evidence:
        injected.append(make_issue(type='SECURITY_RISK', severity='HIGH', layer='L4', source='core', rule_id='AX105', evidence=f'{risk_level_safe} risk with blocking missing evidence', required_fix='Provide evidence'))
        verdict = max_verdict(verdict, 'REVISE')

    # AX501: adjacent confirmation required
    m = DESTRUCTIVE.search(input.output)
    if m and not _has_adjacent_confirmation(input.output, m.start()):
        injected.append(make_issue(type='DATA_RISK', severity='CRITICAL', layer='L4', source='core', rule_id='AX501', evidence=f'Destructive command "{m.group(0)}" without adjacent confirmation', required_fix='Add explicit confirmation requirement near the command'))
        verdict = 'BLOCK'

    # AX502: explicit production deployment only
    if PRODUCTION_DEPLOY.search(input.output) and _risk_idx(risk_level_safe) >= _risk_idx('HIGH'):
        injected.append(make_issue(type='SECURITY_RISK', severity='HIGH', layer='L4', source='core', rule_id='AX502', evidence=f'Production deployment action for {risk_level_safe} risk', required_fix='Require human approval'))
        verdict = max_verdict(verdict, 'ESCALATE_TO_HUMAN')

    # AX503: explicit rollback plan required
    is_migration = input.artifact_type == 'migration' or bool(MIGRATION_RE.search(input.output))
    has_rollback = bool(ROLLBACK_RE.search(input.output)) or bool(ROLLBACK_RE.search(input.evidence.diff))
    if is_migration and not has_rollback:
        injected.append(make_issue(type='DATA_RISK', severity='CRITICAL', layer='L4', source='core', rule_id='AX503', trust_deduction=25, evidence='Migration without explicit rollback plan (requires: rollback plan/script, revert migration, or down migration)', required_fix='Add explicit rollback plan'))
        policy_deductions += 25
        verdict = max_verdict(verdict, 'REVISE')

    # AX504: auth test gap — check type+content, not just rule_id string
    has_auth_test_gap = any(f.rule_id == 'TW205' for f in all_findings) or any(
        f.type == 'TEST_GAP' and f.severity in ('CRITICAL', 'HIGH') and f.evidence and 'auth' in f.evidence.lower()
        for f in all_findings
    )
    if has_auth_test_gap:
        verdict = max_verdict(verdict, 'ESCALATE_TO_HUMAN')

    # CL402
    if any(f.rule_id == 'CL402' for f in all_findings):
        verdict = max_verdict(verdict, 'ESCALATE_TO_HUMAN')

    # LLM signals — guard against non-number confidence
    llm_policy_deductions = 0
    if llm_signals:
        confidence = llm_signals.get('confidence', 1.0)
        if not isinstance(confidence, (int, float)) or isinstance(confidence, bool):
            confidence = 1.0
        if confidence < 0.6 and _risk_idx(risk_level_safe) >= _risk_idx('MEDIUM'):
            injected.append(make_issue(type='AMBIGUITY', severity='MEDIUM', layer='L2', source='llm-overlay', rule_id='CL403', trust_deduction=20, evidence=f'LLM confidence {confidence:.2f} below threshold for {risk_level_safe} risk'))
            llm_policy_deductions += 20
            verdict = max_verdict(verdict, 'ESCALATE_TO_HUMAN')

        # Apply LLM risk_guess if higher
        risk_guess = llm_signals.get('risk_guess', '')
        if risk_guess in RISK_ORDER and _risk_idx(risk_guess) > _risk_idx(risk_level_safe):
            injected.append(make_issue(type='AMBIGUITY', severity='MEDIUM', layer='L2', source='llm-overlay', evidence=f'LLM assessed risk as {risk_guess} vs deterministic {risk_level_safe}', suggested_fix='Review for additional risk signals'))

    # Extension required_verdict: only honor BLOCK from extensions
    has_extension_block = any(
        f.source == 'extension' and getattr(f, 'required_verdict', None) == 'BLOCK' and f.is_blocking
        for f in all_findings
    )
    if has_extension_block:
        verdict = 'BLOCK'

    if verdict == 'BLOCK':
        pass  # terminal

    return {'verdict': verdict, 'injected_issues': injected, 'policy_deductions': policy_deductions + llm_policy_deductions}


def convert_llm_signals(signals: dict) -> list[SealIssue]:
    issues = []
    for field, (issue_type, severity) in LLM_SIGNAL_MAP.items():
        items = signals.get(field)
        if not isinstance(items, list):
            continue
        for text in items:
            if isinstance(text, str) and text.strip():
                issues.append(make_issue(type=issue_type, severity=severity, layer='L2', source='llm-overlay', evidence=text))
    return issues
